---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)

library(tidyverse)

example_input_carbohydrates <- read_tsv("examples/input_carbohydrates_and_degrading_enzymes.tsv")


example_output_carb_degrading_bacteria <- read_tsv("examples/output_prebiotic_degrading_bacteria.tsv")

example_output_bacterial_neuroactive_potential <- read_tsv("examples/output_bacterial_neuroactive_potential.tsv")


```

# KEGG probiotics

<!-- badges: start -->
<!-- badges: end -->

The goal of kegg probiotics is to leverage the data available in the [KEGG database](https://www.genome.jp/kegg/) to make finding bacterial genomes with potential to degrade carbohydrates of interest while metabolising neuroactive compounds an easier task. It was successfully used to produce the *in-silico* analysis conducted in the published article: [Paper name](link to paper).

## Overview

The scripts starts from a list of KEGG compound IDs of carbohydrates (or internal carbohydrate subunits of a polymer), and performs the following steps using the information stored in KEGG:

1.    Connects carbohydrate's compounds IDs to IDs of reactions where it participats
2.    Connects the reaction IDs to Enzymes (EC codes). If EC codes are provided as an additional optional input for some carbohydrates, those EC numbers are integrated into the next step.
3.    Connects EC numbers to Kegg Orthologs (KOs)
4.    Connects KOs to all bacterial genomes where that KO was found.


## Input example

The scripts takes a tabular input data frame in the following form:

```{r show input, echo = FALSE}
knitr::kable(head(example_input_carbohydrates))


```

where:

-  `carbohydrate_name` Identifies the target carbohydrate in an human-friendly way (not used in the script) [optional colum].
-  `carbohydrate_kegg_id` ID of the carbohydrate (or a subunit within a polymer) used in the KEGG db for its identification.
-  `carbohydrate_name` ECs of enzymes known to interact with the carbohydrate of interest [optional column].



## Output example

There are two main outputs produced by the script.

1)    A tabular output with the bacterial genomes showing enzymes capable of degrading the carbohydrates of interest

```{r show carbohydrate output, echo = FALSE}
knitr::kable(head(example_output_carb_degrading_bacteria))

```

2)    A tabular output with the bacterial genomes showing enzymes involved in the metabolism of neuroactive compounds

```{r show neuroactive output, echo = FALSE}
knitr::kable(head(example_output_bacterial_neuroactive_potential))

```



## Comments on the usage of the software

While the script doesn't need the user to provide a list of already known enzymes that can interact with carbohydrates of interest, adding them can increase the performance of the script and make the search of suitable candidates more thorough. Although the curation of such a list is project specific, there are some resources the authors of this repository find useful:

-   Supplementary table 1 from the article [Metatranscriptomic holobiont analysis of carbohydrate-active enzymes in the millipede *Telodeinopus aoutii* (Diplopoda, Spirostreptida)](https://www.frontiersin.org/journals/ecology-and-evolution/articles/10.3389/fevo.2022.931986/full)
-   Supplementary table 1 & 2 from the article [The microbiomeâ€™s fiber degradation profile and its relationship with the host diet](https://bmcbiol.biomedcentral.com/articles/10.1186/s12915-022-01461-6)
-   Fiber_Enzyme_matrix.csv from the [Github](https://github.com/borenstein-lab/IFDP) repository associated with the above-mentioned article.


Finally, the script here provided `find_probiotic_candidates.R` uses the R package [`omixerRpm`](https://github.com/omixer/omixer-rpmR) which is an R interface to the software [Omixer-RPM](https://github.com/raeslab/omixer-rpm), which is distributed under an academic Non-commercial Software License Agreement for Omixer-RPM. Details on the conditions of that licence can be found in the software's [GitHub account](https://github.com/raeslab/omixer-rpm?tab=License-1-ov-file).
